WordPress/gutenberg
# Fix synced pattern editing in write mode and refactor block editing mode to reducer

## What?
Supercedes some previous attempts #66919, #66949, and #65408!
Fixes #66424

In Write Mode, parts of synced patterns that are supposed to be non-editable are editable. This PR fixes the problem, plus undertakes some refactoring of how block editing mode works due to some performance caused by so much logic being in the `getBlockEditingMode` selector.

## Why?
This bug happens because Write Mode overrides the blockEditingMode of a synced pattern's inner blocks.

Pattern blocks try calling `setBlockEditingMode` for child blocks imperatively (in the pattern block's edit function). This does set the correct block editing modes in the block editor store, but when write mode is active the `getBlockEditingMode` selector returns early with a different value.

#65408 tried to fix this by updating the `getBlockEditingMode` selector, but it cause a performance hit. 

## How?
This PR fixes the issue by moving the pattern block editing modes to the block editor store, so that it can work harmoniously with the way Write Mode and Zoom Out mode determine the block editing mode.

The PR also undertakes a refactor of how block editing mode works for zoomed out and write mode to improve performance. The `getBlockEditingMode` select is called a lot (including recursively), so any code added there can cause performance regressions.

The summary of the changes is:
- Remove the zoom out and write mode handling from the `getBlockEditingMode` selector
- Introduce a new `withDerivedBlockEditingModes` higher order reducer, it calculates a derivedBlockEditingMode state for any actions that might require this is recomputed using a few other functions:
    - `getDerivedBlockEditingModesForTree` traverses a block tree calculating the block editing modes for all the blocks in that tree, returning them in a `Map`. This can be used for the entire block tree or just a part of the block tree.
    - `getDerivedBlockEditingModeForBlock` calculates the block editing mode for a single client id. Called by `getDerivedBlockEditingModesForTree` on each block in the tree.
    - `getDerivedBlockEditingModesUpdates` used for actions that add/remove a partial subset of blocks. It should only return a value when there are block editing mode changes, meaning the store state doesn't unnecessarily change.
    - Plus a few utility functions - `traverseBlockTree`, `getBlockTreeBlock`, `findParentInClientIdsList`, `hasBindings`.
    - Also adds a new `isContentBlock` private function to the blocks package.
    
It's quite a lot of code, so I'm open to suggestions on how to structure it.
    
## Testing Instructions
1. Create a synced pattern (Pattern A) with two paragraphs
2. Edit the pattern and enable overrides for one of the patterns
3. Insert the pattern into a template
4. Check that you can only edit the overridden paragraph, the other paragraph should be disabled
5. Enable Zoomed Out mode, you shouldn't be able to edit any of the pattern content, since zoomed out doesn't allow content editing
6. Leave Zoomed Out mode and enable Write mode, check that you can only edit the overridden paragraph in the pattern
7. Create a second pattern (Pattern B) similar to the first one, it should have a block that's overridden and a regular block
8. Edit Pattern A and insert Pattern B into it.
9. Insert Pattern A into a template again, or go back to where you inserted it before
10. Check that you can't edit any of Pattern B from the instance of Pattern A, it should all be disabled

In addition to this, give a good test to both zoomed out and write mode, move blocks around, insert and remove stuff and make sure there are no issues. 😄 

## Screenshots or screencast 
#### Before

https://github.com/user-attachments/assets/49cd5420-aa00-4437-be43-1d1f16002514

#### After

https://github.com/user-attachments/assets/feca05d3-9fb7-404a-8e5a-406e27b4f023
